bolt/lib/Rewrite/RewriteInstance.cpp

```diff
@@ -1276,6 +1280,60 @@ void RewriteInstance::discoverFileObjects() {
   registerFragments();
 }
 
+Error RewriteInstance::discoverFiniAddress() {
+  // If FiniFunctionAddress is already set, we got if from DT_FINI. We use
+  // DT_FINI instead of DT_FINI_ARRAY if it's available.
+  if (BC->FiniFunctionAddress)
+    return Error::success();
+
+  if (BC->IsStaticExecutable)
+    return Error::success();
+
+  if (!BC->FiniArrayAddress || !BC->FiniArraySize) {
+    return createStringError(
+        std::errc::not_supported,
+        "Instrumentation needs either DT_FINI or DT_FINI_ARRAY");
+  }
+
+  if (*BC->FiniArraySize < BC->AsmInfo->getCodePointerSize()) {
+    return createStringError(std::errc::not_supported,
+                             "Need at least 1 DT_FINI_ARRAY slot");
+  }
+
+  ErrorOr<BinarySection &> FiniArraySection =
+      BC->getSectionForAddress(*BC->FiniArrayAddress);
+  if (auto EC = FiniArraySection.getError())
+    return errorCodeToError(EC);
+
+  BC->FiniArraySection = &*FiniArraySection;
+  const Relocation *FiniFunctionReloc =
+      FiniArraySection->getDynamicRelocationAt(0);
+
+  if (!FiniFunctionReloc) {
+    return createStringError(std::errc::not_supported,
```

Comment:
As I understand we're looking for the RELATIVE relocation here. But non-pie binaries won't have one, all we need is emit the right address at the array, so editing static relocations (since instrumentation doesn't support binaries without ones) would be good enough?

Reply:
I belive it should be done in this review. It seems to be not so problematic. For all binaries you should patch static reloc (although RELA relocations doesn't really need this, but RELR needs, and still it would be nice to check the address update using objdump). And for pie binaries also change dynamic reloc. Also it won't hurt to get the first non-pie binary test for risc-v :))

Reply:
> I belive it should be done in this review. It seems to be not so problematic. For all binaries you should patch static reloc

What do you mean exactly with for *all* binaries? For pie binaries, there are no static relocations.

For non-pie binaries, it might indeed not be very difficult to implement this. However, I still have to implement quite a few relocations in order to support non-pie RISC-V binaries so I currently have no way to test this. So I would still propose to keep this for a later patch. Also because this patch is already quite large and usable as is.

Reply:
All binaries have static relocations if they are linked with -Wl,-q, that is what I mean. Even if RISC-V currently not supporting non-pie binaries it is still possible to test with aarch64 test for example quite easily..

Reply:
Updating the static relocations seems a bit trickier than I hoped. The problem is that the runtime fini address is only known *after* `emitAndLink` while the relocations need to be updated *before*.

One way of dealing with this would be to not update the relocation of the first `.fini_array` entry but simply writing the entry itself after `emitAndLink`. This feels slightly hacky though and would also increase the complexity of this patch quite a bit. Do you see another approach?

Reply:
Maybe I'm loosing smth but for static relocation you might just use symbol, not the addend? 
BTW the same thing is possible to do with dynamic relocations too, I have it in my golang support patch, please check the very end if RewriteInstance.cpp https://reviews.llvm.org/D124347 , I add extra logic there so RELATIVE relocations might use symbols too, I think we might use it here so you won't have to patch it using the emitted address. If you'd like I can prepare this part of patch separately

Reply:
> Maybe I'm loosing smth but for static relocation you might just use symbol, not the addend?

The problem is that we don't have an `MCSymbol` for the fini address: we get its value by performing a symbol lookup using the linker after the instrumentation library is loaded (see [here](https://github.com/llvm/llvm-project/blob/f9149a34d9b4ea0a13e83befea6c1b1bcf95b1c4/bolt/lib/RuntimeLibs/InstrumentationRuntimeLibrary.cpp#L212)).

Reply:
I see. Then probably rewriting the address is the only way. Well it seems fine, we're doing it for dynamic relocs, .dynamic & etc, so maybe it is not a big deal to patch also init_array values for runtime library cases..

Reply:
I think I found an elegant way to do this: add a *pending* relocation for the static relocation we want to patch. Since pending relocations are already flushed while rewriting the file, patching the fini address happens automatically.

Note that although I think it's possible to test this on AArch64, I haven't been able to get AArch64 instrumentation working on my x86 machine so I'm not able to verify this. I'll try to fix this later and add tests.